source/extensions/access_loggers/fluentd/fluentd_access_log_impl.h

```diff
@@ -0,0 +1,162 @@
+  COUNTER(entries_buffered)                                                                        \
+  COUNTER(events_sent)                                                                             \
+  COUNTER(connections_closed)
+
+struct AccessLogFluentdStats {
+  ACCESS_LOG_FLUENTD_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT)
+};
+
+class FluentdAccessLoggerImpl : public Tcp::AsyncTcpClientCallbacks,
+                                public FluentdAccessLogger,
+                                public Logger::Loggable<Logger::Id::client> {
+public:
+  FluentdAccessLoggerImpl(Tcp::AsyncTcpClientPtr client, Event::Dispatcher& dispatcher,
+                          const FluentdAccessLogConfig& config, Stats::Scope& parent_scope);
+
+  // Tcp::AsyncTcpClientCallbacks
+  void onEvent(Network::ConnectionEvent event) override;
+  void onAboveWriteBufferHighWatermark() override {}
+  void onBelowWriteBufferLowWatermark() override {}
+  void onData(Buffer::Instance&, bool) override {}
+
+  // FluentdAccessLogger
+  void log(EntryPtr&& entry) override;
+
+private:
+  void flush();
+  void clearBuffer();
+
+  bool disconnected_ = false;
+  bool connecting_ = false;
+  std::string tag_;
+  std::string id_;
+  const Stats::ScopeSharedPtr stats_scope_;
+  AccessLogFluentdStats fluentd_stats_;
+  std::vector<EntryPtr> entries_;
+  uint64_t approximate_message_size_bytes_ = 0;
+  const Tcp::AsyncTcpClientPtr client_;
+  const std::chrono::milliseconds buffer_flush_interval_msec_;
+  const uint64_t max_buffer_size_bytes_;
+  const Event::TimerPtr flush_timer_;
+};
+
+class FluentdAccessLoggerCache {
+public:
+  virtual ~FluentdAccessLoggerCache() = default;
+
+  /**
+   * Get existing logger or create a new one for the given configuration.
+   * @return FluentdAccessLoggerSharedPtr ready for logging requests.
+   */
+  virtual FluentdAccessLoggerSharedPtr
+  getOrCreateLogger(const FluentdAccessLogConfigSharedPtr config) PURE;
+};
+
+using FluentdAccessLoggerCacheSharedPtr = std::shared_ptr<FluentdAccessLoggerCache>;
+
+class FluentdAccessLoggerCacheImpl : public Singleton::Instance, public FluentdAccessLoggerCache {
+public:
+  FluentdAccessLoggerCacheImpl(Upstream::ClusterManager& cluster_manager,
+                               Stats::Scope& parent_scope, ThreadLocal::SlotAllocator& tls);
+
+  FluentdAccessLoggerSharedPtr
+  getOrCreateLogger(const FluentdAccessLogConfigSharedPtr config) override;
+
+private:
+  /**
+   * Per-thread cache.
+   */
+  struct ThreadLocalCache : public ThreadLocal::ThreadLocalObject {
+    ThreadLocalCache(Event::Dispatcher& dispatcher) : dispatcher_(dispatcher) {}
+
+    Event::Dispatcher& dispatcher_;
+    // Access loggers indexed by the hash of logger's configuration.
+    absl::flat_hash_map<std::size_t, FluentdAccessLoggerSharedPtr> access_loggers_;
+  };
+
+  Upstream::ClusterManager& cluster_manager_;
+  const Stats::ScopeSharedPtr stats_scope_;
+  ThreadLocal::SlotPtr tls_slot_;
+};
```

Comment:
Cache is only used on in the main thread, it's unnecessary to use TLS for the cache instance.
It would be better to just store the whole logger (which contains the thread local loggers) in the cache.

And please note the cache should only store the weak_ptr to avoid memory leak.

Reply:
Not sure I understand, because ``FluentdAccessLoggerCacheImpl`` is singleton, but it would create a thread-local cache because, when a new ``FluentdAccessLog`` is created, it posts an operation to create a logger on each worker here:
```
tls_slot_->set(
      [config = config_, access_logger_cache = access_logger_cache_](Event::Dispatcher&) {
        return std::make_shared<ThreadLocalLogger>(access_logger_cache->getOrCreateLogger(config));
      });
```
So eventually this lambda is being invoked on each worker separately, and similarly when ``getOrCreateLogger`` is called, it would pick the cache of that specific worker.

If we won't follow this pattern, we may get the same cache/logger for two different workers (?)
Pls lmk if I got this wrong or if there's another pattern you suggest

Reply:
You can take the tracer cache as an example, the cache will store a single driver object and needn't TLS, the driver object will contains a TLS member.

source/common/tracing/tracer_manager_impl.cc